src/System.Threading.Tasks.Dataflow/src/Blocks/TransformManyBlock.cs

```diff
@@ -538,10 +545,12 @@ private void StoreOutputItemsNonReorderedWithIteration(IEnumerable<TOutput> outp
                 }
                 else
                 {
-                    lock (ParallelSourceLock)
+                    foreach (TOutput item in outputItems)
                     {
-                        foreach (TOutput item in outputItems)
+                        lock (ParallelSourceLock) // don't hold lock while enumerating
+                        {
                             _source.AddMessage(item);
```

Comment:
>_source.AddMessage(item); [](start = 28, length = 25)

These changes looks good. 

I am seeing this lock is used when calling AddMessage. would it be better to add the logic of locking inside SourceCore<TOutput> to just instantiate it with sole flag telling if there is parallelism or not to handle the locking? 

I know this will be bigger change though

Reply:
I agree that would be reasonable to look into (I don't remember whether there's anything that would make that a particularly good or bad change), and I also agree it should be examined separately.